### PR TITLE
refactor: refactor the button component using `forwardRef`

### PR DESCRIPTION
**Pull request by TinsFox**
# Description

refactor the button component using `forwardRef`

The previous button component was disguised as a div, which would lose the original capabilities of HTML, such as form submission. This will prevent users from submitting when JavaScript capabilities are turned off. But when we follow web standards, we can bring a better experience

This submission includes three contents:
1. Use forwardRef to reconstruct the button component
2. Change the original `type` attribute to `variant`. Type is the original attribute of button. We should not modify it.
3. Removed `memo`. As a basic component, memo is not required
Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [x] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement

# How Has This Been Tested?

All button components should behave the same as before

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

### CODE REVIEW

File: web/app/components/base/button/index.tsx [hunk 1]
```diff
@@ -1,50 +1,40 @@
...
 
-  return (
-    <div
-      className={`btn ${style} ${className && className}`}
-      tabIndex={tabIndex}
-      onClick={disabled ? undefined : onClick}
-    >
-      {children}
-      {/* Spinner is hidden when loading is false */}
-      <Spinner loading={loading} className='!text-white !h-3 !w-3 !border-2 !ml-1' />
```
Comment (nite-knite) [inline web/app/components/base/button/index.tsx:45]: Loading state is missing in the refactored component

**Review by nite-knite — CHANGES_REQUESTED** [review]

File: web/app/components/base/button/index.tsx (hunk 1 above)
Comment (TinsFox) [reply web/app/components/base/button/index.tsx:45]: Added it back in commit [`fe05ccf` (#4379)](https://github.com/langgenius/dify/pull/4379/commits/fe05ccf335ffb59e573660678cd2e4d64937bc05)

**Comment by crazywoola** [comment]
@TinsFox Hello there is a conflict in this pr.

**Comment by TinsFox** [comment]
@crazywoola  All conflicts have been resolved.

File: web/app/components/base/button/index.tsx [hunk 2]
```diff
@@ -1,50 +1,44 @@
...
+    variants: {
+      variant: {
+        primary: 'btn-primary disabled:btn-primary-disabled',
+        warning:
+          'btn-warning disabled:btn-warning-disabled',
+        default: 'btn-default disabled:btn-default-disabled',
+      },
+    },
+    defaultVariants: {
+      variant: 'primary',
```
Comment (JzoNgKVO) [inline web/app/components/base/button/index.tsx]:
```suggestion
      variant: 'default',
```
default variant should be 'default'

**Review by JzoNgKVO — CHANGES_REQUESTED** [review]

**Comment by JzoNgKVO** [comment]
@TinsFox 
Compared to the feature `main`, following files should change the original `type` attribute to `variant`

app/components/tools/add-tool-modal/index.tsx
app/components/tools/add-tool-modal/tools.tsx
app/components/tools/provider/detail.tsx
app/components/tools/workflow-tool/index.tsx
app/components/tools/workflow-tool/confirm-modal/index.tsx
app/components/workflow/nodes/parameter-extractor/components/extract-parameter/update.tsx
app/(shareLayout)/webapp-signin/page.tsx

File: web/app/components/base/button/index.tsx (hunk 2 above)
Comment (TinsFox) [reply web/app/components/base/button/index.tsx]: ths!

**Review by JzoNgKVO — APPROVED** [review]

**Review by nite-knite — APPROVED** [review]

**Pull request merged**